### DISCUSSION TITLE
Write out Only a part of the Buffer to the display? (For Cycle efficiency)

### DISCUSSION THREAD

**Discussion by mikelemo** [olikraus/u8g2#1827] [General]
I was just wondering is there is a way to update only a particular part of the Buffer to the display in order to try and save on clock cycles and energy for variables that update frequently that might just take 5% of the display pixels.

is there something like u8g2_sendBuffer_rect(&u8g2, FromX, FromY, ToX, ToY);

**Comment by olikraus** [comment]
There are multiple answers and solutions and probably a combination of them is required to get max speed ;-)

1. setClipWindow https://github.com/olikraus/u8g2/wiki/u8g2reference#setclipwindow: Reduce rendering output. Drawback: Does not reduce data transfer.
2. updateDisplayArea https://github.com/olikraus/u8g2/wiki/u8g2reference#updatedisplayarea: Reduce data transfer. Drawback: Works with tile coordinates, does not work in page mode. 
3. setBufferCurrTileRow and sendBuffer https://github.com/olikraus/u8g2/wiki/u8g2reference#setbuffercurrtilerow: Send one tile row to the display (reduce data transfer). Similar to updateDisplayArea, but works with one complete row and can be used in page mode.

Usually 1 and 2 are combined.

Here is an example for updateDisplayArea: https://github.com/olikraus/u8g2/blob/master/sys/arduino/u8g2_full_buffer/UpdateArea/UpdateArea.ino

Example for clipWindow: https://github.com/olikraus/u8g2/blob/master/sys/arduino/u8g2_page_buffer/ClipWindow/ClipWindow.ino

Example for setBufferCurrTileRow and sendBuffer: https://github.com/olikraus/u8g2/blob/master/sys/arduino/u8g2_page_buffer/UpdatePartly/UpdatePartly.ino

**Reply by mikelemo** [reply]
Thanks for the well documented answer @olikraus I think updateDisplayArea() is the one I'm looking for.
and I was wondering what is missing for is to work in page mode?(Which I Guess is normal graphic pixel work method instead of the usual 8bit height pixel per row of tile mode) I don't mind working on that feature on C.

**Comment by olikraus** [comment]
> I was wondering what is missing for is to work in page mode?

Simple: In page mode only one page is in uC memory, so the memory for the transfer is not there in page mode for updateDisplayArea().
In other words you need to switch to full buffer mode to use updateDisplayArea().

**Reply by mikelemo** [reply]
I got a little confused now, when you say you have to switch to full buffer mode what do you mean? 
I thought the same initialization for the display that works for the normal sendBuffer functionality work for tile mode too.

**Reply by olikraus** [reply]
Page Mode hast a _ 1 _ in the constructor. Full frame mode has _ F _

**Reply by mikelemo** [reply]
Huh that comment clears out some mystery , if it requires migrating the whole Full frame functionality to page mode without an ability to make it complicated I'll pass for now since my system also utilizes non byte binary displaying . 
I thought you could just use it in addition to Full frame mode.

I always wanted to dive into the u8g2 library in depth and contribute to it, and I was wondering if there's some page of info somewhere that explains how say a basic example that print some text and draw shapes on graphic displays jump through the library code and files in an overview and general philosophy for applying and developing hardware and fonts as well as for future functionality support?

**Comment by olikraus** [comment]
> if there's some page of info somewhere

There are some documents, but probably no tutorial for lib developers.

Maybe some basic ideas:
 * u8g2 is based on u8x8
 * There is an overview on u8x8 here: https://github.com/olikraus/u8g2/blob/master/doc/u8g2_software_architecture.png
 * There is a implementation guide for none-Arduino hardware: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform
 * Some information is collected here: https://github.com/olikraus/u8g2/wiki/internal

**Reply by mikelemo** [reply]
OK thanks again for the very useful and understandable answer I'll go over them to get a little more connected with the u8g2 library and hopefully pave contributions to it in the future.    
See you around.
